src/Meshes/thinInstanceMesh.ts

```diff
@@ -228,12 +238,34 @@ Mesh.prototype.thinInstanceBufferUpdated = function(kind: string): void {
     if (kind === "matrix") {
         if (this._thinInstanceDataStorage.matrixBuffer) {
             this._thinInstanceDataStorage.matrixBuffer!.updateDirectly(this._thinInstanceDataStorage.matrixData!, 0, this._thinInstanceDataStorage.instancesCount);
+            this._thinInstanceDataStorage.worldMatrices = null;
         }
     } else if (this._userThinInstanceBuffersStorage?.vertexBuffers[kind]) {
         this._userThinInstanceBuffersStorage.vertexBuffers[kind]!.updateDirectly(this._userThinInstanceBuffersStorage.data[kind], 0);
     }
 };
 
+Mesh.prototype.thinInstanceGetWorldMatrices = function(): Matrix[] {
+    if (!this._thinInstanceDataStorage.matrixData || !this._thinInstanceDataStorage.matrixBuffer) {
+        return [];
+    }
+    const matrixData = this._thinInstanceDataStorage.matrixData;
+
+    if (!this._thinInstanceDataStorage.worldMatrices || this._thinInstanceDataStorage.worldMatrices.length !== this._thinInstanceDataStorage.instancesCount) {
+        this._thinInstanceDataStorage.worldMatrices = new Array<Matrix>();

```

Comment:
~~We could have removed / added a thin instance in sequence, so the check `this._thinInstanceDataStorage.worldMatrices.length !== this._thinInstanceDataStorage.instancesCount` could be false but we would still need to update the `worldMatrices` array. Also, the user can use `thinInstanceSetMatrixAt` that would change a matrix without changing the array size.~~ see next comment

I think the easiest way to handle the "dirtyness" of `worldMatrices` is to reset it to null in: `thinInstanceAdd`, `thinInstanceAddSelf` and `thinInstanceSetMatrixAt`. That way all cases will be handled and you only need the `if (!this._thinInstanceDataStorage.worldMatrices)` part of the check.

Reply:
I went to a live version which is FAR more efficient. Please let me know your thought (If I missed something obvious)

Reply:
Indeed, that's a lot better!

The only thing is that the worldMatrices array will never shrink... Not sure if it is a problem, though.

Reply:
Yeah not a biggy! I'll merge then
Thanks a ton!